benchmark/benchmark-atomic.cpp

```diff
@@ -0,0 +1,223 @@
+/// wrapper.
+template<typename AtomicType, typename Policy>
+RAJA_HOST_DEVICE AtomicType AtomicAdd(AtomicType* acc, const AtomicType val) {
+    return RAJA::atomicAdd(Policy {}, acc, val);
+}
+
+template<typename AtomicType, typename Policy>
+RAJA_HOST_DEVICE AtomicType AtomicMax(AtomicType* acc, const AtomicType val) {
+    return RAJA::atomicMax(Policy {}, acc, val);
+}
+
+/// Function template that allows invoking DESUL atomic with a (int*)(T*, T) signature
+template<typename T, typename Policy, typename DesulAtomicSignature<T>::signature AtomicImpl>
+RAJA_HOST_DEVICE T atomicWrapperDesul(T * acc, T value) {
+    return AtomicImpl(acc, value, raja_default_desul_order{},
+                    raja_default_desul_scope{});
+}
+
+//template<typename T, typename RajaAtomicSignature<T>::signature atomic>
+//class IsDesul : public std::false_type {};
+//
+//template<typename T, typename Policy, typename DesulAtomicSignature<T>::signature AtomicImpl>
+//class IsDesul<atomicWrapperDesul<T, Policy, AtomicImpl>> : public std::true_type {};
+
+
+template<typename AtomicType, typename Policy>
+std::string GetImplName (typename DesulAtomicSignature<AtomicType>::signature) {
+    return "Desul atomic";
+}
+
+template <class ExecPolicy, typename AtomicType, typename RajaAtomicSignature<AtomicType>::signature AtomicImpl, bool test_array = false>
+void TimeAtomicOp(int num_iterations = 2, int array_size = 100) {
+    RAJA::Timer timer;
+
+    for (int i = 0; i < num_iterations; ++i) {
+        AtomicType* device_value = nullptr;
+        int len_array = test_array ? array_size : 1;
+        if (IsGPU<ExecPolicy>::value) {
+            AllocateAtomicDevice(&device_value, len_array);
+        } else {
+            device_value = new AtomicType [len_array];
+        }
+        timer.start();
+        RAJA::forall<ExecPolicy>(RAJA::RangeSegment(0, N),
+        [=] RAJA_HOST_DEVICE(int tid)  {
+            if (test_array) {
+                AtomicImpl(&(device_value[tid % array_size]), 1);
+            } else {
+                AtomicImpl(device_value, 1);
+            }
+        });
+
+        timer.stop();
+        if (IsGPU<ExecPolicy>::value) {
+            DeallocateDeviceAtomic(device_value);
+        } else {
+            delete device_value;
+        }
+
+    }
+
+    double t = timer.elapsed();
+    std::cout << INDENT << INDENT << t << "s" << INDENT;
+    //std::cout << GetImplName(AtomicImpl) << ", ";
+    std::cout << "Number of atomics under contention " << array_size << ", ";
+    std::cout << num_iterations * N << " many atomic operations" << ", ";
+    //std::cout << ExecPolicy::PolicyName();
+    std::cout << std::endl;
+}
+
+int main () {
+    // GPU benchmarks
+    std::cout << "Executing CUDA benchmarks" << std::endl;
+    std::cout << INDENT << "Executing atomic add benchmarks" << std::endl;
+    TimeAtomicOp<ExecPolicyGPU<64>::policy, int, AtomicAdd<int, typename GPUAtomic::policy>, true>(4);
```

Comment:
It may be cleaner and easier to work with to set a constexpr thread block size variable to a default value, such as 256, at the top of the file. I think we agreed in a recent group meeting that thread block size doesn't have a significant performance impact for the simple kernels in this file. @MrBurmark what do you think?

Reply:
That sounds good to me, then we could vary it fairly easily if we wanted to.

Reply:
And we could follow the pattern in RAJA Perf if we wanted to try block size sweeps.

Reply:
I added a for_each_type into RAJA for just these kind of use cases https://github.com/LLNL/RAJA/blob/develop/include/RAJA/util/for_each.hpp#L88